rfc/mediachain-rfc-4.md

```diff
@@ -0,0 +1,328 @@
+ different "sameness" perspectives represented through namespace relationships
+ (e.g. Clarifai can publish a namespace containing only merges based on visual similarity)
+* WKI-based provider records can be reasonably implemented as distributed,
+ freeing orgs from relying on indexers for simple mappings
+
+Works that are mediachain-first and require an identifier will receive one
+from a ticket server/blockchain implementation (TBD). This ID will be
+treated the same way as any other.
+
+### Data Structure Operations
+
+The Phase I data structure interface provided 3 basic operations:
+insert, update, and lookup. The insert method ingested an object into
+the system and returned its canonical identifier. For each identifier,
+the system maintained a chain pointer linking the object to a chain of
+updates. The update method updated the chain pointer by consing a new
+update object on top of the current chain, while lookup retrieved the
+head of the chain for a canonical.
+
+In practice, inserts are idempotent and the chained update model only
+serves to constrain the system by imposing order of operations.
+By dropping ordering and utilizing domain specific identifiers, we can
+eliminate the distinction between insert and update.
+
+The write interface is thus an upsert method, which combines insert
+and update, and adds a statement to the set of objects associated with
+an identifier. If the system has no prior knowledge of the identifier,
+then it creates a new entry in the identifier mapping with a singleton
+set containing the object.
+
+The lookup method can be retained as a basic read operation, with a
+change at its return value. Since there are no chained updates any
+more, the method returns the known set of updates relating to an
+identifier. 
+
+### Namespace Partitioning
+
+In order to reflect the natural organization of data according to institution, industry, topic, etc, we introduce the concept of namespaces, which immediately give us several advantages. In terms of scaling, the partitioning must be semantic so as to allow peers to commit resources only in topics they are interested in. In terms of administrative control, the partitioning must be hierarchical so that we can easily delegate management and moderation to organizational stake holders, see "Moderation and Publishing Model" below.
+
+To satisfy these requirements, we've chosen to adopt a hierarchical structure similar to unix paths. Leaves in the hierarchy map to specific datasets and provide the primary entry points for writes, and namespaces higher in the hierarchy aggregate namespaces below and provide a progressively expanded read view of the data structure. Much like symlinks with unix paths, alternate views of the hierarchy are possible, namespaces may have multiple parents, and so on. This creates a semantically rich and expressive framework for querying and composition, see "Namespace-based Queries and Aggregation" below.
+
+### Moderation and Publishing Model
```

Comment:
Would it be correct to say that moderated namespaces assume non-byzantine behavior? I don't think that's the actual assumption but I'm not sure how to put it more precisely (trying to contrast to "open" namespaces which may have a coin fee or similar to write)